Issue 8719: Add basic support for :active selector

sjmelia opened the pull request:
<!-- Please describe your changes on the following line: -->

Added toggling of active state for element and parents on mousedown/mouseup. Active state is removed when mouseout. (hover)
- As with my other PR i'm struggling a bit with the automated testing. I've added a manual test case and found quirks-mode/active-and-hover-manual.html which - aside from also being a manual test, is functional in Firefox but does not render correctly in Servo.
- Not implemented: In Firefox, behaviour differs with a <!DOCTYPE HTML> and an anchor does not lose it's activation on mouseout; whereas a button does.

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #8719  (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/11781)

<!-- Reviewable:end -->

sjmelia committed:
Issue 8719: Add basic support for :active selector

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @mbrubeck (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/document.rs, components/script/dom/activation.rs, components/script/dom/window.rs, components/script/dom/element.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

sjmelia commented:
Can I leave comments on the github diff? I should say that I fought the borrow checker and lost during the DOM traversal so I just recursed - I anticipate needing to revisit this and replace with a loop if i'm heading in the right direction.

emilio commented:
@sjmelia sure! Feel free to leave any comments in the diff. You should be able to use reviewable too.

Sorry this is taking so long to review. The approach does look good to me, but I'm not confident enough with this part of the code to sign off it.

Maybe @mbrubeck, @Ms2ger or @jdm can review this soon-ish?

emilio commented on components/script/dom/activation.rs:
```diff
@@ -29,6 +32,25 @@ pub trait Activatable {
+
+        let win = window_from_node(self.as_element());
+        win.reflow(ReflowGoal::ForDisplay,
+                       ReflowQueryType::NoQuery,
```
Nit: align this with the first paren:

```rust
win.reflow(ReflowGoal::ForDisplay,
           ReflowQueryType::NoQuery, ...)

emilio commented on components/script/dom/activation.rs:
```diff
@@ -29,6 +32,25 @@ pub trait Activatable {
+
+        let win = window_from_node(self.as_element());
+        win.reflow(ReflowGoal::ForDisplay,
+                       ReflowQueryType::NoQuery,
```
ditto

emilio commented:
Let's see if there are any tests for this btw.

@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 0d1ae24 with merge 48a7efe...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1880)

emilio commented:
It seems you need to fix some build errors (probably needing a rebase), since there are a few types that have been moved around :)

Ms2ger commented:
I think our resident activation expert is...

r? @Manishearth

Manishearth commented:
Overall this looks good, but it doesn't seem to handle button inputs -- the formal activation state will be entered even if they are disabled. Perhaps we should override enter_formal_activation_state for inputs?

Or, better, add a should_enter_formal_activation method on the activation trait, have it return true for most elements, and for button inputs, `<a>`, `<link>`, you return true only if they aren't disabled/have href/etc.

sjmelia commented:
Congratulations on starting nightlies!

I've rebased and fixed the nits; regarding the button input the formal activation state is not entered when they're disabled; because Element.as_maybe_activatable (see diff) will return None according to rules which are encapsulated in the elements themselves e.g. htmlinputelement.is_instance_activatable

I've added some disabled/link/a elements to my manual test case to allow for demonstrating this; Firefox (47) still gives them the selector, presumably because that is not specified by CSS3.

sjmelia commented:
Sorry just realised that i'm not quite right/precise there; the method `is_instance_activatable` is on the `Activatable` trait; with a different implementation for each element type. I.e. the logic is not really smeared around on the different elements as I mistakenly implied above; it's a lot more similar to your suggestion in that `should_enter_formal_activation` is synonymous with `is_instance_activatable`

Manishearth commented:
Makes sense.

@bors-servo r+

bors-servo commented:
:pushpin: Commit 421c354 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 421c354d44f93ff8f52affde1b595f4e14b9ed27 -->

bors-servo commented:
:hourglass: Testing commit 421c354 with merge 5afdf7f...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1808), [arm32](http://build.servo.org/builders/arm32/builds/1927), [arm64](http://build.servo.org/builders/arm64/builds/1919), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1958), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2074), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1993), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1981), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2049), [windows](http://build.servo.org/builders/windows/builds/1658)
